### PR TITLE
Use pty instead of an output channel

### LINKED ISSUE #56: Output channel is noticeably delayed
It probably waits a few moments before flushing the line. The delay could be problematic when watching logs

### PR DESCRIPTION
**Pull request by SeaDve**
Fixes #56 and #33

Also make Command::run non async. 

Also renames FlatpakTerminal -> FlatpakRunner. Since the terminal outputting on it is detangled

### CODE REVIEW

File: src/extension.ts [hunk 1]
```diff
@@ -34,9 +35,15 @@ export async function activate(context: ExtensionContext): Promise<void> {
     // Mark the app as already initialized
     store.manifestSelected(manifest)
 
-    const outputChannel = window.createOutputChannel('Flatpak')
-    // Create a Flatpak pty
-    const terminal = new FlatpakTerminal(outputChannel)
+    const terminal = new FlatpakTerminal()
+    const runner = new FlatpakRunner()
+
+    terminal.onDidClose(() => {
```
Comment (SeaDve) [inline src/extension.ts:41]: This may create reference cycle, not sure how typescript handles memory

Comment (bilelmoussaoui) [reply src/extension.ts:41]: I don't think typescript has some way to deal with ref cycles

**Comment by SeaDve** [comment]
ansi color codes works on a pseudoterminal, its probably got escaped somewhere in the line

**Comment by bilelmoussaoui** [comment]
> ansi color codes works on a pseudoterminal, its probably got escaped somewhere in the line

Something that might potentially affects this https://github.com/bilelmoussaoui/flatpak-vscode/blob/master/src/utils.ts#L33

**Comment by SeaDve** [comment]
colors outputted from the actual app somehow has colors 
image

**Comment by bilelmoussaoui** [comment]
Maybe we need to pass the same env variables I mentioned above when we spawn a new child process; see https://nodejs.org/api/child_process.html#child_processspawncommand-args-options

**Comment by SeaDve** [comment]
another very weird thing i noticed, every output that is supposed to be colored is outputted from stderr instead of stdout

**Comment by bilelmoussaoui** [comment]
> another very weird thing i noticed, every output that is supposed to be colored is outputted from stderr instead of stdout
> 
> I might be wrong

I will play a bit  with plain nodejs API a bit and see what is going on

**Comment by bilelmoussaoui** [comment]
Ok I think I found out what is going on, the issue comes actually from the `child_process.spawn` API, see https://stackoverflow.com/questions/7725809/preserve-color-when-executing-child-process-spawn. From the suggested options, the `node-pty` suggestion seems to work pretty fine
```javascript
var pty = require('node-pty');
process.env.FORCE_COLOR = true;
const ps = pty.spawn('flatpak-builder', [
    'repo',
    '--force-clean',
    '/var/home/bilelmoussaoui/Projects/Authenticator/build-aux/com.belmoussaoui.Authenticator.Devel.json'
], { env: process.env });

ps.on('data', (input) => {
    process.stdout.write(`\r${input}\r`)
});
```

**Comment by SeaDve** [comment]
aah. I used pretty_env_logger colored logs to check if coloring works and env_logger by default outputs them on stderr

**Comment by SeaDve** [comment]
Got this error when the extension launches `The module '/home/dave/Projects/flatpak-vscode/node_modules/node-pty/build/Release/pty.node' was compiled against a different Node.js version using NODE_MODULE_VERSION 93. This version of Node.js requires NODE_MODULE_VERSION 89.`

**Comment by bilelmoussaoui** [comment]
Special treatment like what exactly?

**Comment by SeaDve** [comment]
https://github.com/microsoft/vscode/issues/84439

**Comment by bilelmoussaoui** [comment]
The suggestion in the comment does "work" though?

**Comment by SeaDve** [comment]
i haven't got it to work

**Comment by SeaDve** [comment]
finally got it to work, but how it is imported is very ugly.

**Comment by bilelmoussaoui** [comment]
> finally got node-pty to work, but how it is imported is very ugly.
> 
> it works really well right now, and I think the ugliness is worth it

That sounds good to me as a compromise

File: src/flatpakTerminal.ts [hunk 2]
```diff
@@ -1,113 +1,57 @@
...
 
-  onError(message: string, command: Command): void {
-    this.writeError(message)
-    failure({ command, message })
-    this.failed = true
-    this.isRunning = false
-  }
+    appendMessage(message: string, isErr: boolean): void {
+        const boldRed = '\x1b[1;31m'
+        const boldWhite = '\x1b[1;37m'
```
Comment (bilelmoussaoui) [inline src/flatpakTerminal.ts:30]: Why bold by default?

File: src/nodePty.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import * as vscode from 'vscode'
```
Comment (bilelmoussaoui) [inline src/nodePty.ts:1]: This file should get a small comment that at least point to the issue where you got the code snippet

File: src/flatpakTerminal.ts (hunk 2 above)
Comment (bilelmoussaoui) [reply src/flatpakTerminal.ts:30]: Well I am not sure whether this is used only for displaying the executed command or is it used for all the outputs that are not errors

Comment (SeaDve) [reply src/flatpakTerminal.ts:30]: appendMessage is used for the executed command and errors. The output of the process is directed to it untampered

**Comment by bilelmoussaoui** [comment]
Otherwise looks pretty good! thanks for the awesome contributions

File: src/flatpakTerminal.ts (hunk 2 above)
Comment (SeaDve) [reply src/flatpakTerminal.ts:30]: `Nothing to do` also uses appendMessage. What about `appendError` and `appendMessage`?

Comment (bilelmoussaoui) [reply src/flatpakTerminal.ts:30]: Sounds good

**Comment by SeaDve** [comment]
Currently, the behavior of deleting the terminal is killing the child process/cancelling build. Would it be better if it wouldn't end the process and create a new terminal if a new line/output is received?

**Comment by bilelmoussaoui** [comment]
Right, that is problematic. What I would suggest here is to check when the terminal is closed (there is an event for that) and re-create & hide it. Then we can just re-show it again. 

But if someone closes the terminal, that is on them. It is better to stop the build in that case

**Comment by SeaDve** [comment]
> Right, that is problematic. What I would suggest here is to check when the terminal is closed (there is an event for that) and re-create & hide it. Then we can just re-show it again.

This is the current behavior, except when the terminal is closed, it won't be recreated until the show() method is called (something like on demand showing)

**Comment by bilelmoussaoui** [comment]
Then this sounds good to me!

**Comment by bilelmoussaoui** [comment]
I suppose we are good to get this merged?

**Comment by SeaDve** [comment]
show-output-channel may also need a rename?

**Comment by bilelmoussaoui** [comment]
I would keep it and rename it instead, people don't necessarily have to use it but it is there in case someone is not very familiar with VSCode

**Comment by SeaDve** [comment]
I think its ready

**Pull request merged**